autodocodec/src/Autodocodec/Codec.hs

```diff
@@ -156,6 +159,30 @@ data Codec context input output where
     -- | Codec for the 'Right' side
     Codec context input2 output2 ->
     Codec context (Either input1 input2) (Either output1 output2)
+  -- | Encode/decode a discriminated union of objects
+  --
+  -- The type of object being encoded/decoded is discriminated by
+  -- a designated "discriminator" property on the object which takes a string value.
+  --
+  -- When encoding, the provided function is applied to the input to obtain a new encoder
+  -- for the input. The function 'mapToEncoder' is provided to assist with building these
+  -- encoders. See examples in 'Usage.hs'.
+  --
+  -- When decoding, the value of the discriminator property is looked up in the `HashMap`
+  -- to obtain a decoder for the output. The function `mapToDecoder' is provided
+  -- to assist with building these decoders. See examples in 'Usage.hs'.
+  --
+  -- The 'HashMap' is also used to generate schemas for the type.
+  -- In particular, for OpenAPI 3, it will generate a schema with a 'discriminator', as defined
+  -- by https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/
+  DiscriminatedUnionCodec ::
+    -- | propertyName to use for discrimination
+    Text ->
+    -- | how to encode the input
+    (input -> (Discriminator, ObjectCodec input ())) ->
+    -- | how to decode the output
+    HashMap Discriminator (Text, ObjectCodec Void output) ->
```

Comment:
I don't understand yet what this `Text` is for. Could you please elaborate on this?
If it's just to do a `named` or a `<?>` then it should maybe not be here.

Reply:
Sorry, this was documented as part of `SomeDecodable`, which I've since removed. It's the name of the object type, used for generating the OpenAPI3 schema reference. There are two reasons I didn't just rely on using `named` instead:
1. the type name is required (although I guess we could default to using the discriminator value if it's not provided)
2. `named` does not allow naming `ObjectCodec`s, only `ValueCodec`s.

Reply:
I'll add some documentation

Reply:
Done

Reply:
Aw how annoying that that is necessary. Such an ugly feature.